state/endpoint_bindings.go

```diff
@@ -446,22 +421,58 @@ func DefaultEndpointBindingsForCharm(charmMeta *charm.Meta) map[string]string {
 	return bindings
 }
 
-// translateSpaceNameToID takes a map of endpoint bindings to space names and
-// and returns a map of endpoint bindings to space ids
-func (st *State) translateSpaceNameToID(current map[string]string) (map[string]string, error) {
-	retVal := make(map[string]string, len(current))
-	namesToIDs, err := st.SpaceIDsByName()
+type EndpointBinding interface {
+	SpaceByID(id string) (*Space, error)
+	SpaceIDsByName() (map[string]string, error)
+	SpaceNamesByID() (map[string]string, error)
+}
+
+type Bindings struct {
+	st EndpointBinding
+	bindingsMap
+}
+
+// NewBindings returns a bindings guaranteed to be in space id format.
+func NewBindings(st EndpointBinding, givenMap map[string]string) (*Bindings, error) {
+	spacesNamesToIDs, err := st.SpaceIDsByName()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	spacesIDsToNames, err := st.SpaceNamesByID()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	newMap := make(map[string]string, len(givenMap))
+	for endpoint, space := range givenMap {
+		if id, haveName := spacesNamesToIDs[space]; haveName {
```

Comment:
A caveat here is that there might be some ambiguity here if users name their spaces 1,2,3... Is it a spaceID or a spaceName?

How do you feel about being more explicit by creating two constructors, e.g. NewBindingsForApplication(app) and NewBindingsFromSpaceNames(st, givenMap)?

Reply:
Working on EndpointBindings() returning a Binding object to replace the first constructor.  Can't compeletely do away with it because of the bundle export code using EndpointBindings() from the description package, and the upgrade code.  I will be strengthening the constructor based on the concern over a space name looking like a space id.